scipy/stats/_mannwhitneyu.py

```diff
@@ -35,143 +50,100 @@ def pmf(self, k, m, n):
-        shape_new = np.array((m+1, n+1, k+1))
-        if np.any(shape_new > shape_old):
-            shape = np.maximum(shape_old, shape_new)
-            fmnks = -np.ones(shape)             # create the new array
-            m0, n0, k0 = shape_old
-            fmnks[:m0, :n0, :k0] = self._fmnks  # copy remembered values
-            self._fmnks = fmnks
-
-    def _f(self, m, n, k):
-        '''Recursive implementation of function of [3] Theorem 2.5'''
-
-        # [3] Theorem 2.5 Line 1
-        if k < 0 or m < 0 or n < 0 or k > m*n:
-            return 0
-
-        # if already calculated, return the value
-        if self._fmnks[m, n, k] >= 0:
-            return self._fmnks[m, n, k]
-
-        if k == 0 and m >= 0 and n >= 0:  # [3] Theorem 2.5 Line 2
-            fmnk = 1
-        else:   # [3] Theorem 2.5 Line 3 / Equation 3
-            fmnk = self._f(m-1, n, k-n) + self._f(m, n-1, k)
-
-        self._fmnks[m, n, k] = fmnk  # remember result
-
-        return fmnk
-
-
-# Maintain state for faster repeat calls to mannwhitneyu w/ method='exact'
-_mwu_state = _MWU()
-
-
-def _mwu_f_iterative(m, n, k, fmnks):
-    '''Iterative implementation of function of [3] Theorem 2.5'''
-
-    def _base_case(m, n, k):
-        '''Base cases from recursive version'''
-
-        # if already calculated, return the value
-        if fmnks.get((m, n, k), -1) >= 0:
-            return fmnks[(m, n, k)]
-
-        # [3] Theorem 2.5 Line 1
-        elif k < 0 or m < 0 or n < 0 or k > m*n:
-            return 0
-
-        # [3] Theorem 2.5 Line 2
-        elif k == 0 and m >= 0 and n >= 0:
-            return 1
-
-        return None
-
-    stack = [(m, n, k)]
-    fmnk = None
-
-    while stack:
-        # Popping only if necessary would save a tiny bit of time, but NWI.
-        m, n, k = stack.pop()
-
-        # If we're at a base case, continue (stack unwinds)
-        fmnk = _base_case(m, n, k)
-        if fmnk is not None:
-            fmnks[(m, n, k)] = fmnk
-            continue
-
-        # If both terms are base cases, continue (stack unwinds)
-        f1 = _base_case(m-1, n, k-n)
-        f2 = _base_case(m, n-1, k)
-        if f1 is not None and f2 is not None:
-            # [3] Theorem 2.5 Line 3 / Equation 3
-            fmnk = f1 + f2
-            fmnks[(m, n, k)] = fmnk
-            continue
-
-        # recurse deeper
-        stack.append((m, n, k))
-        if f1 is None:
-            stack.append((m-1, n, k-n))
-        if f2 is None:
-            stack.append((m, n-1, k))
-
-    return fmnks
+    # build_sigma_array and build_u_freqs_array adapted from code
+    # by @toobaz with permission. Thanks to @andreasloe for the suggestion.
+    # See https://github.com/scipy/scipy/pull/4933#issuecomment-1898082691
+    def build_sigma_array(self, a):
+        n1, n2 = self.n1, self.n2
+        if a + 1 <= self.s_array.size:
+            return self.s_array[1:a+1]
+
+        s_array = np.zeros(a + 1, dtype=int)
```

Comment:
Alternatively, since we're caching the null distribution for a given `(n1, n2)`, we could make `s_array` and `configurations` as large as they would ever need to be (`int(n1 * n2)`, I think). Then, if the function were called with a higher value of the statistic, only the new values would need to be calculated and added in. This can be left to a follow-up.